dash_auth/oidc_auth.py

```diff
@@ -0,0 +1,288 @@
+        )
+        return kwargs
+
+    def login_request(self):
+        """Login the user."""
+        kwargs = {"_external": True}
+        if self.force_https_callback:
+            kwargs["_scheme"] = "https"
+        if request.headers.get("X-Forwarded-Host"):
+            host = request.headers.get("X-Forwarded-Host")
+            redirect_uri = f"https://{os.path.join(host, self.callback_route)}"
+        else:
+            redirect_uri = url_for("oidc_callback", **kwargs)
+
+        oauth_client = self.oauth_client
+        if oauth_client is None:
+            if self.idp_selection_redirect:
+                return redirect(self.idp_selection_redirect)
+            return (
+                "Several OAuth providers are registered. Please choose one.",
+                400,
+            )
+        session["idp_name"] = self.idp_name
+
+        return oauth_client.authorize_redirect(
+            redirect_uri,
+            **self.oauth_kwargs.get("authorize_redirect_kwargs", {}),
+        )
+
+    def logout(self):  # pylint: disable=C0116
+        """Logout the user."""
+        session.clear()
+        base_url = self.app.config.get("url_base_pathname") or "/"
+        page = self.logout_page or f"""
+        <div style="display: flex; flex-direction: column;
+        gap: 0.75rem; padding: 3rem 5rem;">
+            <div>Logged out successfully</div>
+            <div><a href="{base_url}">Go back</a></div>
+        </div>
+        """
+        return page, 200
+
+    def callback(self):  # pylint: disable=C0116
+        """Do the OIDC dance."""
+        oauth_client = self.oauth_client
+        oauth_kwargs = self.oauth_kwargs
+        del session["idp_name"]
+        if oauth_client is None:
+            return (
+                "Several OAuth providers are registered. Please choose one.",
+                400,
+            )
+
+        try:
+            token = oauth_client.authorize_access_token(
+                **oauth_kwargs.get("authorize_token_kwargs", {}),
+            )
+        except OAuthError as err:
+            return str(err), 401
+        user = token.get("userinfo")
+        if user:
+            session["user"] = user
+            if "offline_access" in oauth_client.client_kwargs["scope"]:
+                refresh_token = token.get("refresh_token")
+                session["refresh_token"] = refresh_token
+            if self.log_signins:
+                logging.info("User %s is logging in.", user.get("email"))
+
+        return redirect(self.app.config.get("url_base_pathname") or "/")
+
+    def is_authorized(self):  # pylint: disable=C0116
+        """Check whether ther user is authenticated."""
+        return (
+            request.path in [
+                self.login_route,
+                self.logout_route,
+                self.callback_route,
+                self.idp_selection_redirect,
+            ]
+            or "user" in session
+        )
+
+
+def get_oauth(app: dash.Dash = None) -> OAuth:
```

Comment:
I see. would setting a `oidcauth` attribute / property or even method on `app` be more discoverable?

Reply:
I don't think so, you would still need to either import app or dash's `get_app` to retrieve it, then get the attribute from it, without intellisense since this attribute doesn't exist on the base Dash class.

Reply:
OK. fair.